org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensVisitor.java

```diff
@@ -499,23 +485,111 @@ public boolean visit(TypeDeclaration node) {
 		return false;
 	}
 
-	private static IScanner getScanner() {
-		if (fScanner == null) {
-			fScanner = ToolFactory.createScanner(true, false, false, true);
+	/**
+	 * Tries to create an {@link IScanner} for the source of the given compilation unit.
+	 *
+	 * @param cu the compilation unit
+	 * @return the scanner, or {@code null} if not available
+	 */
+	private IScanner createScanner(CompilationUnit cu) {
+		final ITypeRoot typeRoot = cu.getTypeRoot();
+		final IJavaProject javaProject = typeRoot.getJavaProject();
+		if (typeRoot == null || javaProject == null) {
+			return null;
+		}
+		final String source;
+		try {
+			source = typeRoot.getSource();
+		} catch (JavaModelException e) {
+			return null;
+		}
+		if (source == null) {
+			return null;
 		}
-		return fScanner;
+
+		final String sourceLevel = javaProject.getOption(JavaCore.COMPILER_SOURCE, true);
+		final String complianceLevel = javaProject.getOption(JavaCore.COMPILER_COMPLIANCE, true);
+		final boolean enablePreview = JavaCore.ENABLED.equals(javaProject.getOption(JavaCore.COMPILER_PB_ENABLE_PREVIEW_FEATURES, true));
+
+		final IScanner scanner = ToolFactory.createScanner(false, false, false, sourceLevel, complianceLevel, enablePreview);
+		scanner.setSource(source.toCharArray());
+		return scanner;
 	}
 
-	private int getNextToken(IScanner scanner) {
-		int token = 0;
-		while (token == 0) {
+	private int getNextValidToken(IScanner scanner) {
+		while (true) {
 			try {
-				token = scanner.getNextToken();
+				return scanner.getNextToken();
 			} catch (InvalidInputException e) {
 				// ignore
 			}
 		}
-		return token;
+	}
+
+	/**
+	 * Visitor for {@link IScanner} tokens.
+	 */
+	@FunctionalInterface
+	private interface ScannerTokenVisitor {
+		/**
+		 * Visits the given scanner token.
+		 *
+		 * @param scannerToken the scanner token ID, see {@link ITerminalSymbols}
+		 * @param tokenOffset the document offset of the scanner token
+		 * @param tokenLength the length of the scanner token
+		 */
+		public void visit(int scannerToken, int tokenOffset, int tokenLength);
+	}
+
+	/**
+	 * Uses an {@link IScanner} (if available) to tokenize a source range in the document,
+	 * and visits the scanner tokens in order of occurence in the source range.
+	 *
+	 * <p>
+	 *     <strong>NOTE:</strong> If semantic tokens are added by the visitor, the scan range MUST NOT intersect
+	 *     with any other range where semantic tokens can appear, in order to avoid overlapping semantic tokens.
+	 * </p>
+	 *
+	 * @param startPosition the (inclusive) start position of the scan range
+	 * @param endPosition the (exclusive) end position of the scan range
+	 * @param tokenVisitor the visitor to use for scanner tokens
+	 */
+	private void tokenizeWithScanner(int startPosition, int endPosition, ScannerTokenVisitor tokenVisitor) {
+		if (scanner == null) {
+			return;
+		}
+		scanner.resetTo(startPosition, endPosition - 1); // -1 because resetTo wants inclusive endPosition
```

Comment:
I think this is somewhat subjective. My preference is to use `[A, B)` intervals (inclusive start, exclusive end), for a number of reasons ([nicely summarized here](https://stackoverflow.com/a/9810359)).

> you're passing in the start position of another element as opposed to the correct ending position

The start position of one element is (conveniently) the end position of another (adjacent) element in the `[A, B)` world.

> Wouldn't it make more sense to change that line to `int gapBeforeNameEnd = typeDeclaration.getName().getStartPosition() - 1;` ?

I could, if I decided to live in the `[A, B]` world, but then I will have to introduce a bunch of `+1`'s and/or `-1`'s everywhere, which I find annoying (as it can lead to off-by-one errors). One of the nice things about the `[A, B)` world is that `start + length = end`, as opposed to the `[A, B]` world where `start + length = end - 1`. The only time you need to do `+1` or `-1` in the `[A, B)` world is when working with an API that doesn't use `[A, B)` intervals (e.g. `IScanner`).

I think (without having done much research at all) that `[A, B)` is very common, probably for the advantages listed above. For example, `java.lang.String#substring(int beginIndex, int endIndex)` uses `[A, B)` intervals.

Anyways, these were just my thoughts :slightly_smiling_face:. I don't mind changing the code to use `[A, B]` intervals if you prefer it.

Reply:
Ok, let's leave it as-is then :stuck_out_tongue_winking_eye: . It's not that I prefer having the end index be inclusive, but I felt that as `tokenizeWithScanner(..)` is a helper method for the scanner, it should respect the same convention for those that might be familiar with the API. With that said, the javadocs make it clear it's exclusive.

Maybe it's better that those making the nice contributions decide what convention they prefer :)